clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -206,3 +208,124 @@ void SemaSYCL::handleKernelEntryPointAttr(Decl *D, const ParsedAttr &AL) {
   D->addAttr(::new (SemaRef.Context)
                  SYCLKernelEntryPointAttr(SemaRef.Context, AL, TSI));
 }
+
+static SourceLocation SourceLocationForType(QualType QT) {
+  SourceLocation Loc;
+  const Type *T = QT->getUnqualifiedDesugaredType();
+  if (const TagType *TT = dyn_cast<TagType>(T))
+    Loc = TT->getDecl()->getLocation();
+  else if (const ObjCInterfaceType *ObjCIT = dyn_cast<ObjCInterfaceType>(T))
+    Loc = ObjCIT->getDecl()->getLocation();
+  return Loc;
```

Comment:
Should there be an 'assert' in the 'else' here?  It seems like a lot of types here are missing!

Reply:
No, it is intentional that an invalid source location be returned if there is no source location for the type (e.g., for `int`). I think the above covers all (non-alias) types that can be user declared. I could add comments to better explain the intent for this function; I'll do that.

Reply:
I renamed the function to `SourceLocationForUserDeclaredType()` and added comments to explain its intended purpose.